android/src/main/java/com/reactNativeQuickActions/AppShortcutsModule.kt

```diff
@@ -0,0 +1,170 @@
+import androidx.core.net.toUri
+import com.facebook.react.bridge.ActivityEventListener
+import com.facebook.react.bridge.Callback
+import com.facebook.react.bridge.JSApplicationIllegalArgumentException
+import com.facebook.react.bridge.Promise
+import com.facebook.react.bridge.ReactApplicationContext
+import com.facebook.react.bridge.ReactContextBaseJavaModule
+import com.facebook.react.bridge.ReactMethod
+import com.facebook.react.bridge.ReadableArray
+import com.facebook.react.bridge.WritableMap
+import com.facebook.react.module.annotations.ReactModule
+import com.facebook.react.modules.core.DeviceEventManagerModule
+import com.reactNativeQuickActions.ShortcutItem.Companion.fromReadableArray
+
+@ReactModule(name = AppShortcutsModule.REACT_NAME)
+internal class AppShortcutsModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
+    init {
+        reactContext.addActivityEventListener(object : ActivityEventListener {
+            override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
+                // Do nothing
+            }
+
+            override fun onNewIntent(intent: Intent) {
+                sendJSEvent(intent)
+            }
+        })
+    }
+
+    override fun getName(): String {
+        return REACT_NAME
+    }
+
+    @ReactMethod
+    @TargetApi(25)
+    fun popInitialAction(promise: Promise) {
+        try {
+            val currentActivity = currentActivity
+            var map: WritableMap? = null
+            if (currentActivity != null) {
+                val intent = currentActivity.intent
+                if (ACTION_SHORTCUT == intent.action) {
+                    val shortcutItem: ShortcutItem? = IntentCompat.getParcelableExtra(intent, SHORTCUT_ITEM, ShortcutItem::class.java)
+                    if (shortcutItem != null) {
+                        map = shortcutItem.toWritableMap()
+                    }
+                }
+            }
+            promise.resolve(map)
+        } catch (e: Exception) {
+            promise.reject(
+                JSApplicationIllegalArgumentException(
+                    "AppShortcuts.popInitialAction error. " + e.message
+                )
+            )
+        }
+    }
+
+    @ReactMethod
+    @TargetApi(25)
+    fun setShortcutItems(items: ReadableArray) {
+        if (!isShortcutSupported || items.size() == 0) {
+            return
+        }
+        val currentActivity = currentActivity ?: return
+
+        val shortcutItems = fromReadableArray(items)
+        val shortcuts = shortcutItems.mapIndexed { index: Int, item: ShortcutItem ->
+            return@mapIndexed createShortcutInfo(currentActivity, id = "id$index", item)
+        }
+
+        try {
+            ShortcutManagerCompat.setDynamicShortcuts(reactApplicationContext, shortcuts)
+        } catch (error: Throwable) {
+            Log.e(LOG_TAG, "Error while setting dynamic shortcuts.", error)
```

Comment:
Error is gracefully handled, but not returned like in `popInitialAction`

Reply:
True, but the way the library was initially set up, `setShortcutItems()` does not use any callback/promise so there is no way (yet) to return a result.

The thing that I changed here is to simply adding try-catch to prevent exceptions which would cause the react-native app to crash.

But I totally agree that we should add a Promise here further down the road. But this would also mean to change the iOS and typescript layer, which is not feasible for now.